src/ripple/server/impl/BaseWSPeer.h

```diff
@@ -214,7 +215,8 @@ send(std::shared_ptr<WSMsg> w)
     {
         JLOG(this->j_.info()) <<
             "closing slow client";
-        cr_.code = static_cast<boost::beast::websocket::close_code>(4000);
+        cr_.code = safe_cast<boost::beast::websocket::close_code>
+                               (static_cast<unsigned short>(4000u));
```

Comment:
I think this is overkill. In fact, I think we could probably swap this to be:

    cr_.code = boost::beast::websocket::close_code::policy_error;
    cr_.reason = "Policy error: client is too slow.";

Reply:
Is the change of value from 4000 (not a `close_code`) to 1008 (`close_code::policy_error`) intentional?  I'm not sure of the meaning of 4000, and was leery of changing behavior.

The type of `cr_.code` is `uint16_t`, and the type of `close_code::policy_error` is `enum close_code : std::uint16_t`, so this would be an ideal case for `safe_cast` (which lives to do the enum <-> integral conversions safely):

```
cr_.code = safe_cast<std::uint16_t>(boost::beast::websocket::close_code::policy_error);
```

or:

```
cr_.code = safe_cast<decltype(cr_.code)>(boost::beast::websocket::close_code::policy_error);
```

Reply:
In this situation we _are_  actually closing the connection, so a close code is appropriate.

Looking at the WebSocket spec, codes in the 4000 range are "_reserved for private use and thus can't be registered.  Such codes can be used by prior agreements between WebSocket applications.  The interpretation of these codes is undefined by this protocol._" ([RFC 6455 7.4.2](https://tools.ietf.org/html/rfc6455#section-7.4.2))

The "policy error" message seemed appropriate (the "policy" in question being "no slow clients") however, RFC 6455 says that 1008 "_indicates that an endpoint is terminating the connection because it has *received a message* that violates its policy._" (emphasis mine).

However, we are in the "write" codepath here; we haven't "received" something, so I'm unsure if this is an appropriate error code to use. The RFC does say that 1008 "_is a generic status code that can be returned when there is no other more suitable status code_" but I'm no longer sure that it makes sense.

I'm fine with leaving this bit as is.

Reply:
Analysis of theoretical close codes is always fun but in this case we should be pragmatic. None of the clients connecting to rippled's WebSocket interfaces are going to know what "4000" means. But they will all know what 1007 (`policy_error`) means.